### PR TITLE
[4.0] Fields/editor/codemirror as custom element

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
Move the inline jQuery dependant code to custom elements/native ES6 javascript

### Testing Instructions
Install this branch, change the default editor to codemirror and check it's functionality (F10 and all the highlighting still works)

screen shot 2018-03-12 at 01 11 37
screen shot 2018-03-12 at 01 11 58

### Few notes
- All the related scripts (codemirror, addons the custom element itself) are lazy loaded
- thus it makes this editor PWA ready
- We're using Promises (the polyfill for the webcomponents cover the older browsers)
- we're using requestAnimationFrame (https://caniuse.com/#feat=requestanimationframe) supported even in IE11 (wow 😮)
- the code remains flexible as the options can be changed in runtime (even better than the old var approach)

### Documentation Changes Required
Yes

### CODE REVIEW

**Comment by infograf768** [comment]
Does not look like working here (installed from your branch):
screen shot 2018-03-12 at 09 42 18

**Comment by infograf768** [comment]
Loading failed for the <script> with source “http://localhost:8888/joomla-cms-fields-editor-codemirror/joomla-cms-fields-editor-codemirror/media/vendor/codemirror/lib/codemirror.min.js”.

Looks like domain is added twice

**Comment by dgrammatiko** [comment]
Thanks will check it

**Comment by infograf768** [comment]
Same for addons.js as I tested by adding in index.php
```
HTMLHelper::_('script', 'media/vendor/codemirror/lib/codemirror.js');
HTMLHelper::_('script', 'media/vendor/codemirror/lib/addons.js');
```
which still get the error in console but loads the textarea normally.

Also, you still have some debug code in FileLayout.php

**Comment by dgrammatiko** [comment]
And a slight facelift:
screen shot 2018-03-12 at 12 42 48

**Comment by anuragteapot** [comment]
screenshot from 2018-03-12 16-57-24

@dgt41 not fixed yet.

**Comment by dgrammatiko** [comment]
@infograf768 @Anu1601CS since I never use subfolders to install any instance of Joomla can you please provide the error messages and a screenshot of the DOM element `joomla-editor-codemirror` ?
Thanks

**Comment by anuragteapot** [comment]
screenshot from 2018-03-13 01-25-20
screenshot from 2018-03-13 01-24-57

**Comment by dgrammatiko** [comment]
@Anu1601CS let me know if latest commit solves the problem

**Comment by infograf768** [comment]
Not working yet. Looks related to the slash
```
‘src’ attribute of <script> element is not a valid URI: “http://localhost:8888:8888\joomla-cms-fields-editor-codemirror/media/vendor/codemirror/lib/codemirror.min.js”
```
(can you Update branch?)

**Comment by dgrammatiko** [comment]
@infograf768 can you fetch/pull and try again?

**Comment by infograf768** [comment]
Loading correctly now.
Reminder: You still have some debug code in FileLayout.php

**Comment by dgrammatiko** [comment]
> You still have some debug code in FileLayout.php

What do you mean? This PR is not touching any FileLayout.php

**Comment by infograf768** [comment]
maybe I did not express myself correctly.
When debug is on, I get this. But indeed it looks like it is voluntary and independant from your PR.

screen shot 2018-03-13 at 09 53 58

**Comment by dgrammatiko** [comment]
@infograf768 that's the debug code for JLayouts, seems related to the PR but it shouldn't be. Do you get the same on the main repo?

**Comment by infograf768** [comment]
Yep, I do. No p.

**Comment by anuragteapot** [comment]
@dgt41 Now its little bit fine.

screenshot from 2018-03-13 16-04-17
:) But some bug in height of input pane.

**Comment by dgrammatiko** [comment]
@Anu1601CS more a css thing: https://stackoverflow.com/a/44954280

Ps the answer is the one below that one I shared (stupid mobile layout)

**Comment by C-Lodder** [comment]
@dgt41 - Please put the XTD button back below the textarea. The textarea is the most important part of the edit view, so when those buttons stack on mobile, users will have to scroll down even further.

**Comment by dgrammatiko** [comment]
The buttons will become dropdown similar to tinymce for all editors but cannot be done yet as we don’t have the dropdown custom element yet...
Also the buttons need a new API...
So please hold on here as this is the starting point, there are few more steps to get to the final version

**Comment by C-Lodder** [comment]
Dropdown or not, prioritization of content. Simple UX.

**Comment by brianteeman** [comment]
When there is no wysiwyg toolbar the buttons should be at the bottom as they are less important than the editor

**Comment by dgrammatiko** [comment]
Then according to this tinyMCE should get the button (all of them) in the bottom. Seriously now I try to make this consistent in all 3 core editors, but if people think that only tiny should get the xtd-buttons on the top toolbar and the rest should have a different ux/ui then that’s fine for me, let’s keep confusing people. Joomla is well known for this 😂

**Comment by C-Lodder** [comment]
Well if I were up to me, the TinyMCE toolbar would also be placed at the bottom too.

That said, people don't really switch between editors. They stick with the editor that suites them best. So if they're used to using Codemirror and having XTD buttons that the bottom, leave them there

**Comment by dgrammatiko** [comment]
> So if they're used to using Codemirror and having XTD buttons that the bottom, leave them there

In that sense let's revert the left menu to top because that is what ppl are used to? 
I will always try to make things consistent, now if others don't care or think that this is a bad idea and it's acceptable to keep broken UX/UI, then fine. Will revert it back the 10 years old approach...

**Comment by C-Lodder** [comment]
Dimitris, is a simple matter of prioritizing content. We did research on this area ages ago.

The sidebar menu is there because it fixes the issue in J3 where too many components is problematic. It actually fixes something, not just a random thing I initially did for the sake of it.

Please explain why having buttons on the bottom o the editor is a "10 year old approach"...I'm interested.

**Comment by infograf768** [comment]
I guess the best is to have the choice between top and bottom.

Examples: 
I always have to scroll in frontend when saving an article because the Save buttons are at the bottom while I would prefer them on top.
Another one is when installing Joomla and choosing to install languages. Instead of having the Buttons on top AND bottom, they are only at the bottom, forcing to scroll that long list when the language you want to install is towards the top.

**Comment by dgrammatiko** [comment]
> I always have to scroll in frontend when saving an article because the Save buttons are at the bottom while I would prefer them on top.

Thats the reason in every other software the buttons **are always on top**

> forcing to scroll that long list when the language you want to install is towards the top.

The problem there is that you have a very long page which in fact should be only one multiple selection field, so no scrolling at all.

Anyways, as I said before this is about consistency, but people like to have a messy Joomla so I'll revert these changes, np. Let's keep the bad UX which is the trademark of Joomla

**Comment by dgrammatiko** [comment]
Reverted, enjoy the broken UX

**Comment by brianteeman** [comment]
If only we had a UX team who could do some user testing

**Comment by dgrammatiko** [comment]
@infograf768 @Anu1601CS can you please mark your tests here?

**Comment by laoneo** [comment]
Would be good to have some tests here.

File: plugins/editors/codemirror/layouts/editors/codemirror/element.php [hunk 1]
```diff
@@ -10,31 +10,40 @@
...
-	jQuery(function () {
-		var id = ' . json_encode($id) . ', options = ' . json_encode($options) . ';
-		/** Register Editor */
-		Joomla.editors.instances[id] = CodeMirror.fromTextArea(document.getElementById(id), options);
-	});
-');
-?>
+HTMLHelper::_('stylesheet', $basePath . 'lib/codemirror' . $extCSS, array('version' => 'auto'));
+HTMLHelper::_('stylesheet', $basePath . 'lib/addons' . $extCSS, array('version' => 'auto'));
+HTMLHelper::_('webcomponent', ['joomla-editor-codemirror' => 'system/webcomponents/joomla-editor-codemirror.js'], array('version' => 'auto', 'relative' => true));
```
Comment (anuragteapot) [inline plugins/editors/codemirror/layouts/editors/codemirror/element.php]: @dgt41 please fix this.  And use minified version. Thanks

**Comment by anuragteapot** [comment]
After I manually fix above issue and when i tried  to install in subfolder
screenshot from 2018-04-05 18-13-37

**Comment by anuragteapot** [comment]
Error 
Loading failed for the <script> with source “http://localhost/anurag/joomla-cms-fields-editor-codemirror/anurag/joomla-cms-fields-editor-codemirror/media/vendor/codemirror/lib/codemirror.min.js”

**Comment by brianteeman** [comment]
@dgt41 you really need to switch your dev environment so that you are developing in a sub folder all this will keep happening again and again ;)

**Comment by dgrammatiko** [comment]
The commit https://github.com/joomla/joomla-cms/pull/19890/commits/a5b1f7ab6768c4ace346fe4155d9f14e6c6b06be is totally wrong I will revert it

**Comment by dgrammatiko** [comment]
@Anu1601CS can you confirm that the editor is ok now?

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on 95e949ff4f0b6ec305e4ee3dbbb4aeba9301ccc0<br /><br />@dgrammatiko Now it's fine for me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19890">issues.joomla.org/tracker/joomla-cms/19890</a>.</sub>

**Comment by zero-24** [comment]
@laoneo please make sure you assign prs to milestones ;-)

**Pull request merged**